File: 29_Craps/javascript/craps.js
// CRAPS
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// Main program
async function main()
{
    print(tab(33) + "CRAPS\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    r = 0;
    print("2,3,12 ARE LOSERS: 4,5,6,8,9,10 ARE POINTS: 7,11 ARE NATURAL WINNERS.\n");
    t = 1;
    print("PICK A NUMBER AND INPUT TO ROLL DICE");
    z = parseInt(await input());
    do {
        x = Math.random();
        t++;
    } while (t <= z) ;
    while (1) {
        print("INPUT THE AMOUNT OF YOUR WAGER.");
        f = parseInt(await input());
        print("I WILL NOW THROW THE DICE\n");
        do {
            e = Math.floor(7 * Math.random());
            s = Math.floor(7 * Math.random());
            x = e + s;
        } while (x == 0 || x == 1) ;
        if (x == 7 || x == 11) {
            print(x + " - NATURAL....A WINNER!!!!\n");
            print(x + " PAYS EVEN MONEY, YOU WIN " + f + " DOLLARS\n");
            r += f;
        } else if (x == 2) {
            print(x + " - SNAKE EYES....YOU LOSE.\n");
            print("YOU LOSE " + f + " DOLLARS.\n");
            r -= f;
        } else if (x == 3 || x == 12) { // Original duplicates comparison in line 70
            print(x + " - CRAPS....YOU LOSE.\n");
            print("YOU LOSE " + f + " DOLLARS.\n");
            r -= f;
        } else {
            print(x + " IS THE POINT. I WILL ROLL AGAIN\n");
            while (1) {
                do {
                    h = Math.floor(7 * Math.random());
                    q = Math.floor(7 * Math.random());
                    o = h + q;
                } while (o == 0 || o == 1) ;
                if (o == 7) {
                    print(o + " - CRAPS, YOU LOSE.\n");
                    print("YOU LOSE $" + f + "\n");
                    r -= f;
                    break;
                }
                if (o == x) {
                    print(x + " - A WINNER.........CONGRATS!!!!!!!!\n");
                    print(x + " AT 2 TO 1 ODDS PAYS YOU...LET ME SEE..." + 2 * f + " DOLLARS\n");
                    r += f * 2;
                    break;
                }
                print(o + " - NO POINT. I WILL ROLL AGAIN\n");
            }
        }
        print(" IF YOU WANT TO PLAY AGAIN PRINT 5 IF NOT PRINT 2");
        m = parseInt(await input());
        if (r < 0) {
            print("YOU ARE NOW UNDER $" + -r + "\n");
        } else if (r > 0) {
            print("YOU ARE NOW AHEAD $" + r + "\n");
        } else {
            print("YOU ARE NOW EVEN AT 0\n");
        }
        if (m != 5)
            break;
    }
    if (r < 0) {
        print("TOO BAD, YOU ARE IN THE HOLE. COME AGAIN.\n");
    } else if (r > 0) {
        print("CONGRATULATIONS---YOU CAME OUT A WINNER. COME AGAIN!\n");
    } else {
        print("CONGRATULATIONS---YOU CAME OUT EVEN, NOT BAD FOR AN AMATEUR\n");
    }
    
}

main();

File: 29_Craps/ruby/craps.rb
class CRAPSGAME

    # class variables start with a double "@"
    @@standings = 0

    def displayHeading
        puts "CRAPS".center(80)
        puts "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY".center(80)
        puts "\n\n\n"
        puts "2,3,12 are losers"
        puts "4,5,6,8,9,10 are points"
        puts "7,11 are natural winners.\n\n"
    end

    def displayStanding
        if @@standings < 0
            print "you are in the hole by "
        elsif @@standings == 0
            print "you currently have "
        else
            # show how much money we currently have
            print "you now have won "
        end
           # print the absolute value of the amount in the standings
        puts @@standings.abs.to_s + " dollars"
    end

    # dice can come up 2 through 12
    # so return a minimum of 2 and add 0 through 10 to that
    def rollDice
        puts "I will now throw the dice"
        return rand(5) + rand(5) + 2
    end

    def placeBet
        print "How much do you want to wager? "
        wager = gets.strip.to_i
        return wager
    end

    def loseBetBy amount
        @@standings -= amount
    end

    def winBetBy amount
        @@standings += amount
    end

    def askQuit?
        print "\nDo you want to play again? "
        # just the first character, make it uppercase
        again = gets.strip.upcase[0]
        return again != "Y"
    end

    def pointRoll point, wager
        while true do
            puts " is the point."
            puts " I will roll again when you press Enter."
            waitForIt = gets
            roll = rollDice
            print roll.to_s

            # the only critical rolls here are 7 and the previous roll
            # if anything else comes up we roll again.
            case roll.to_i
                when 7
                    puts " craps - you lose"
                    loseBetBy wager
                    break
                when point
                    puts " is a winner! congrats!"
                    puts "at 2 to 1 odds pays you " + (2 * wager).to_s + " dollars"
                    winBetBy 2 * wager
                    break
                else
                    print " no point - " + point.to_s
           end
        end
    end

    def play
        displayHeading
        
        while true do
            wagerAmount = placeBet
            roll = rollDice
            print roll.to_s
            case roll
                when 2
                    puts " snake eyes - you lose"
                    loseBetBy wagerAmount
                when 3, 12
                    puts " craps - you lose"
                    loseBetBy wagerAmount
                when 4, 5, 6, 8, 9, 10
                    pointRoll roll, wagerAmount
                when 7, 11
                    puts " a natural - a winner"
                    puts "pays even money: " + wagerAmount.to_s + " dollars"
                    winBetBy wagerAmount
            end
            displayStanding
            if askQuit?
                endPlay
            end
        end
    end

    def endPlay
        case
            when @@standings < 0
                puts "Too bad. You are in the hole " + @@standings.abs.to_s + " dollars. Come again."
            when @@standings > 0
                puts "Congratulations --- You came out a winner of " + @@standings.to_s + " dollars. Come again!"
            when @@standings == 0
                puts "Congratulations --- You came out even, not bad for an amateur"
        end
        exit
    end
end

craps = CRAPSGAME.new
craps.play

